docs/plugins/testing_workshop_docs/article-3-readers-and-fixtures.md

```diff
@@ -0,0 +1,152 @@
+* Article 4: [Test coverage](./article-4-test-coverage.md)  
+* Resource links: [Testing resources](./testing-resources.md)  
+
+### This article covers:   
+* [Readers](#reader)  
+* [Built-in fixtures](#built-in-fixtures)  
+* [Custom fixtures and round-trip tests](#custom-fixtures-and-round-trip-tests)  
+* [Enclosed testing](#enclosed-testing)  
+  
+### Resources  
+The example plugin and all the tests discussed in this article are available in [this GitHub repository](https://github.com/DragaDoncila/plugin-tests).  
+  
+## Introduction  
+In this article, we discuss a plugin called `plugin_tests`, generated using the cookiecutter, which has a reader and a widget. The reader is the cookiecutter numpy file reader, `napari_get_reader`. It checks whether a path ends in `.npy`. If it doesn't, it returns `None`, and if it does, it returns the `reader_function`, which loads the data. 
+
+![napari_get_reader](../../images/napari_plugins_1st_napari_get_reader.PNG)
+  
+## Reader
+The `napari_get_reader` function is the first thing to test. In the top-level directory under `src`, we have the `plugin_tests` module. Inside `plugin_tests` is the `_tests` directory. This is a typical structure when writing tests. There is also a `test_reader.py` file, which is empty. We will populate it with tests.  
+
+![reader_function](../../images/napari_plugins_2nd_reader_function.png)
+  
+Test as much as possible and focus on writing small tests that look at one indivisible unit. We are focused on testing the `napari_get_reader` function. Sometimes it returns `None`; sometimes it returns the `reader_function`. We want to ensure that if we pass in a path that ends with `.npy`, it gives us back a function we can call.  
+  
+We import `numpy` and `napari_get_reader`. `numpy` will be used later.  
+  
+Best practice: Give tests meaningful names that describe what they're doing.
+  
+## Built-in fixtures  
+We use `tmp_path` to manage the writing and reading of files during test execution. `tmp_path` is a `pytest` fixture; it is not imported, it comes with `pytest`. We pass `tmp_path` as a parameter to our test function, and `pytest` will inject it when the tests run. `tmp_path` provides a temporary path used to save and manipulate files. Temporary paths, files, and directories created in this way during the testing process are automatically removed by `pytest` when the tests are completed. 
+  
+We create a file to read. First, we’ll build a file path, then generate a small amount of data. We’re just testing to see if the function returns a `callable` as expected, so we don't need a large array.
+  
+There are no specific requirements for the contents of the array in this case. We just need some sort of file to save to this temporary directory. The test file will not appear anywhere unless there is a pause during test execution.   
+  
+`temp_dir` is another fixture. It provides a full working directory. `.mkdir` and other commands can be used inside `temp_dir`.
```

Comment:
```suggestion
```

temp_dir (I assume tmpdir is intended) is not really recommended for use anymore, so I wouldn't mention it (see https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html#the-tmp-path-fixture)